rfcs/20220307-monitoring_across_the_stack.md

```diff
@@ -0,0 +1,355 @@
+
+Prometheus configurations can be linted by `promtool` using this pre-commit repo: https://github.com/fortman/pre-commit-prometheus.
+
+## Outstanding questions
+
+### Authentication
+
+Do we want to hide the metrics endpoint behind authentication? I don't really see why we would other than to stop spamming (but we could throttle those endpoints to match the prometheus scrape rate for example and it would prevent anyone from doing too much damage).
+
+I like the idea of having fully open Prometheus and Grafana dashboards that don't require authentication to _explore_. Obviously we'd want to hide administration inside of authentication but each Prometheus and Grafana handle that on their own.
+
+### Alerting
+
+Do we want to use Prometheus for alerting or Grafana?
+
+Pros for Prometheus:
+- It is possible and potentially even trivial to provision alerts based on configuration files
+
+Cons for Prometheus:
+- It is much harder to configure and automatically provision changes to Prometheus's configuration
+- The UI for building them does not exist so I'm not sure what the best way to write new rules is other than via yaml files which will be harder for folks to contribute to given we're not all full-time devops people with tons of time to learn how to do this without fiddling around with visual helper tools
+
+Pros for Grafana:
+- The UI for building them is nice
+
+Cons for Grafana:
+- There's no way to export or provision them so the only way to add rules is through the UI (there's no JSON version of the form either). Moving rules across environments would be hard... to be fair, developing rules with anything except production data is also very difficult (lots of guessing) so maybe this isn't a problem.
+
+I'd rather use Grafana provided we back up the database of alert configurations. If we do that then not having them in the code is hopefully less of a disaster.
+
+It is possible to export _all_ of them via the rules API: `http://grafana/api/ruler/grafana/api/v1/rules`. There's just no way to upload that JSON blob directly. We could scrape the endpoint and back it up into the code base for good measure and to further distribute the configuration.
+
+### Updating configurations
+
+#### Grafana
+
+Grafana is configured to automatically detect new dashboards uploaded via JSON. It needs a restart to add new datasources and notification channels via the `provisioning` folder[^3].
+
+There is currently no way to export or provision alerting rules.
+
+#### Prometheus
+
+Prometheus is configured via a single `prometheus.yml` file. I've created a Python script that generatese the appropriate `prometheus.{environment}.yml` file based on a `.conf` file using Python's `configparser` module. It's in the contents of this PR.
+
+## Overview of proposed changes
+
+* Use ECS to deploy the following services in production:
+    * Prometheus
+    * Grafana
```

Comment:
Note that Grafana is already deployed in our production infrastructure. I'm not sure if we'll replace or reuse the existing one. Probably whichever is less work (presumably reusing).

cc @zackkrida for final decision there.

Reply:
The current service is so simple, we should just be able to update the grafana version and redepoy. That's what I recommend. We can look at switching to ECS later.